### PR TITLE
Simplify the EventBasedLRUCache and Allows Skipping LRU

### LINKED ISSUE #1782: Simplify the LRU Cache
This task is to simplify the EventBaseLRUCache a bit and improve the performance by adding a new flag that skips the LRU maintenance of the _double LinkedList as not everyone needs it (especially if the cache is never full and removeExpiredValues is false).

### PR DESCRIPTION
**Pull request by dannybtsai**
This task is to simplify the EventBaseLRUCache a bit and improve the performance by adding a new flag that skips the LRU maintenance of the _double LinkedList as not everyone needs it (especially if the cache is never full and removeExpiredValues is false).

### CODE REVIEW

File: src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs [hunk 1]
```diff
@@ -136,24 +102,43 @@ private int EventQueuePollingInterval
...
             IEqualityComparer<TKey> comparer = null,
             bool removeExpiredValues = false,
-            int removeExpiredValuesIntervalInSeconds = 300)
+            int removeExpiredValuesIntervalInSeconds = 300,
+            bool maintainLRU = false)
         {
             _capacity = capacity > 0 ? capacity : throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException(nameof(capacity)));
             _options = options;
             _map = new ConcurrentDictionary<TKey, LRUCacheItem<TKey, TValue>>(comparer ?? EqualityComparer<TKey>.Default);
             _removeExpiredValuesIntervalInSeconds = 1000 * removeExpiredValuesIntervalInSeconds;
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs]: why do we multiple by 1000 here?

File: src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs [hunk 2]
```diff
@@ -272,19 +251,6 @@ private void RemoveLRUs()
             }
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs:352]: Could we write: _doubleLinkedList.RemoveLast()?

File: src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs [hunk 3]
```diff
@@ -136,24 +102,43 @@ private int EventQueuePollingInterval
...
+            _maintainLRU = maintainLRU;
+            _dueForExpiredValuesRemoval = DateTime.UtcNow.AddSeconds(_removeExpiredValuesIntervalInSeconds);
         }
 
+        /// <summary>
+        /// Occurs when the application is ready to exit.
+        /// </summary>
+        /// <param name="sender">The sender of the event.</param>
+        /// <param name="e">The event argument.</param>
+        private void DomainProcessExit(object sender, EventArgs e) => StopEventQueueTask();
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs]: How are these methods used?

**Review by brentschmaltz — APPROVED** [review]

**Comment by pmaytak** [comment]
@dannybtsai 
What is the difference operationally between this EventBasedLRUCache with maintainLRU = false and MaximumSizeCache in #1771? The way I udnerstand it - when compacting MaximumSizeCache removes entries randomly from the dictionary (there's no linked list), while this one removes expired entries, and also when compacting it removes entries based on the linked list? (Although the LRU order of the list is not maintained?)


@pmaytak , what you described is correct and the idea in MaximumSizeCache is to provide a simple/lightweight, non-LRU cache that does not have Linkedlist. So, maintaining a Linkedlist is the only difference.

File: src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs [hunk 4]
```diff
@@ -307,11 +333,14 @@ public bool SetValue(TKey key, TValue value, DateTime expirationTime)
             {
                 cacheItem.Value = value;
                 cacheItem.ExpirationTime = expirationTime;
-                AddActionToEventQueue(() =>
+                if (_maintainLRU)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs:404]: @dannybtsai I am not understanding how the map is compacted.
The method RemoveLRUs() checks the _doubleLinkedList.Count>0 on line 291, but i don't see where an item is added to _doubleLinkedList ?

**Review by brentschmaltz — COMMENTED** [review]
Trying to understand how the cache gets cleared out

File: src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs (hunk 4 above)
Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs:404]: Danny and i reviewed this and it seems fine, i just needed some hand holding....

However, we decided, that is we are using the MRU part, just skip adding to the _doubleLinkList and modify the RemoveLRU to only consider the _doubleLinkList if MRU logic is in.

We also suggested to rename RemoveLRU to Compact.

File: src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs [hunk 5]
```diff
@@ -347,21 +415,26 @@ public bool SetValue(TKey key, TValue value, DateTime expirationTime)
                 // if cache is at _maxCapacityPercentage, trim it by _compactionPercentage
                 if ((double)_map.Count / _capacity >= _maxCapacityPercentage)
                 {
-                    _eventQueue.Enqueue(() =>
+                    if (_maintainLRU)
+                        _eventQueue.Enqueue(CompactLRU);
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs:419]: nice to create separate methods for LRU and Non-LRU

**Review by brentschmaltz — APPROVED** [review]

**Pull request merged**